src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -487,16 +533,11 @@ protected virtual void CheckProjectCanBeRemoved(ProjectId projectId)
         {
         }
 
-        private void HandleProjectChange(ProjectId projectId, Func<Solution, Solution> getSolutionWithChangedProject)
+        private void HandleProjectChange(ProjectId projectId, Func<Solution, Solution> solutionTransformation)
         {
-            using (_serializationLock.DisposableWait())
+            if (SetCurrentSolutionWithLock(solutionTransformation, out var oldSolution, out var newSolution))
             {
-                CheckProjectIsInCurrentSolution(projectId);
-
-                var oldSolution = this.CurrentSolution;
-                var newSolution = this.SetCurrentSolution(getSolutionWithChangedProject(oldSolution));
-
-                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectChanged, oldSolution, newSolution, projectId);
+                RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectChanged, oldSolution, newSolution, projectId);
```

Comment:
Isn't this going to have ordering issues where events can get out of order with the setting now? Because SetCurrentSolutionWithLock on two threads could complete, and then the RaiseWorkspaceChangedEventAsync could run in a different order?

Reply:
Yes, I'll move the event back under the lock.

---
In reply to: [392485165](https://github.com/dotnet/roslyn/pull/42228#discussion_r392485165) [](ancestors = 392485165)